byzcoin/bcadmin/lib/utils.go

```diff
@@ -70,3 +75,43 @@ func GetDarcByID(cl *byzcoin.Client, id []byte) (*darc.Darc, error) {
 
 	return d, nil
 }
+
+// AddTransactionWithOption add an option to export the transaction instread of
+// sending it. An export will redirect the transaction to stdout.
+func AddTransactionWithOption(c *cli.Context, cl *byzcoin.Client, tx byzcoin.ClientTransaction) (*byzcoin.AddTxResponse, error) {
+	return AddTransactionAndWaitWithOption(c, cl, tx, 0)
+}
+
+// AddTransactionAndWaitWithOption adds an option to export the transaction
+// instead of sending it. An export will redirect the transaction to stdout.
+func AddTransactionAndWaitWithOption(c *cli.Context, cl *byzcoin.Client, tx byzcoin.ClientTransaction, wait int) (*byzcoin.AddTxResponse, error) {
+	export := c.GlobalBool("export")
+
+	if export {
+		// When exporting, we must not pass SignerCounter, SignerIdentities and
+		// Signatures. Hence, we build a new list of instructions by ommiting
+		// those parameters. We can't edit current ones because those are not
+		// pointers.
+		instrs := make([]byzcoin.Instruction, len(tx.Instructions))
+		for i, instr := range tx.Instructions {
+			instrs[i] = byzcoin.Instruction{
+				InstanceID: instr.InstanceID,
+				Spawn:      instr.Spawn,
+				Invoke:     instr.Invoke,
+				Delete:     instr.Delete,
+			}
+		}
+		tx.Instructions = instrs
+		buf, err := protobuf.Encode(&tx)
+		if err != nil {
+			return nil, errors.New("failed to encode tx: " + err.Error())
+		}
+		reader := bytes.NewReader(buf)
+		_, err = io.Copy(os.Stdout, reader)
+		if err != nil {
+			return nil, errors.New("failed to copy to stdout: " + err.Error())
+		}
+		os.Exit(0)
+	}
+	return cl.AddTransactionAndWait(tx, wait)
```

Comment:
the way the functions are arranged in this file is a bit strange.

I think it'll be cleaner if there are two distinct functions. one for adding a real transaction, the other for printing the transaction. 

Something like this:
```
TransactionToString(cl *byzcoin.Client, tx byzcoin.ClientTransaction) (string, error)
AddTransaction(...)
```

actually you don't need the second one because you can just do `cl.AddTransactionAndWait` or `cl.AddTransaction`.

The caller needs to decide which one to use.

Reply:
the way I see it is `AddTransactionAndWaitWithOption` does two very different things depending on the global flag. there is no overlap between the two tasks (the `if` statement covers most of the function) except the parameters. so there's no need to put these two functionalities into one function.

Reply:
The downside of it is that we must then implement the if/else block everywhere we commit a transaction (because we want it to be widely available): 

```go
if c.GlobalBool("export") {
  TransactionToString(...)
} else {
  AddTransaction(...)
}
```

...or I could implement this if/else block in a new function function here? `AddWaitOrPrintTransaction(...)`?

Reply:
personally i prefer seeing the explicit `if/else`, it's less "magical"

Reply:
I agree. @jeffallen is it okay like that?